keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,503 @@
+  - [Existing controller](#existing-controller)
+  - [New controller](#new-controller)
+    - [Informers and Caches](#informers-and-caches)
+    - [Multiple workers](#multiple-workers)
+    - [Handling Cron aspect](#handling-cron-aspect)
+  - [Metrics](#metrics)
+  - [Add .status.lastSuccessfulTime](#add-statuslastsuccessfultime)
+  - [Fix applicable open issues](#fix-applicable-open-issues)
+  - [Scale Targets for GA](#scale-targets-for-ga)
+    - [CronJob Limits](#cronjob-limits)
+    - [Frequency of launched jobs](#frequency-of-launched-jobs)
+- [API changes](#api-changes)
+  - [CronJob v1 API](#cronjob-v1-api)
+  - [CronJob v1beta1 API](#cronjob-v1beta1-api)
+  - [Validations](#validations)
+- [Tests](#tests)
+  - [E2E test](#e2e-test)
+    - [Existing test cases](#existing-test-cases)
+    - [New test cases](#new-test-cases)
+  - [Conformance Tests](#conformance-tests)
+  - [Unit Tests](#unit-tests)
+- [Implementation plan](#implementation-plan)
+    - [Release X](#release-x)
+    - [Release X+1](#release-x1)
+    - [Release X+2](#release-x2)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Alternatives and Further Reading](#alternatives-and-further-reading)
+    - [Cron Aspect](#cron-aspect)
+- [Improvements/Considerations](#improvementsconsiderations)
+  - [Add .status.nextScheduleTime](#add-statusnextscheduletime)
+  - [Add counters](#add-counters)
+  - [Add .status.conditions](#add-statusconditions)
+  - [Support Jitter for cronjobs](#support-jitter-for-cronjobs)
+  - [Support Timezone for cronjobs](#support-timezone-for-cronjobs)
+  - [CronJob umbrella issue](#cronjob-umbrella-issue)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [x] "Implementation History" section is up-to-date for milestone
+- [x] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [x] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+[CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is a Kubernetes API that creates Job object on a schedule specified by a cron spec. It is in beta status since v1.8 and currently it is marked deprecated with target release for removal being v1.22. This document lays out the plan to promote it to stable.
```

Comment:
Sorry - recovering my previous comment about removal date.

From the KEP:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/1635-prevent-permabeta#proposal

```
This means that every beta API will be deprecated in nine months and removed in 18 months.
```

So doesn't that mean that deprecation will happen in 1.22 and removal in 1.25?

@deads2k - can you please chime in too?

Reply:
Deprecation in 1.22, it will be unservable in 1.25.  However, if the API doesn't make it to GA in 1.22 or before, then another beta is required before going GA.  See https://github.com/kubernetes/enhancements/blame/master/keps/sig-architecture/1635-prevent-permabeta/README.md#L94

I think we'd really like to avoid double churn.

Reply:
Thanks David - apparently I missed that line.

And sure - I fully agree we really should do everything we can to avoid double churn.

Reply:
Updated per above comment.